tests/unit/ViolationUtilsTest.js

```diff
@@ -0,0 +1,192 @@
+    name: 'missingTag',
+    type: 'violation',
+    userMessage: '',
+};
+
+describe('getViolationsOnyxData', () => {
+    let transaction;
+    let transactionViolations;
+    let policyRequiresTags;
+    let policyTags;
+    let policyRequiresCategories;
+    let policyCategories;
+
+    beforeEach(() => {
+        transaction = {transactionID: '123'};
+        transactionViolations = [];
+        policyRequiresTags = false;
+        policyTags = {};
+        policyRequiresCategories = false;
+        policyCategories = {};
+    });
+
+    it('should return an object with correct shape', () => {
+        const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
+        expect(result).toEqual({
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transaction.transactionID}`,
+            value: transactionViolations,
+        });
+    });
+
+    it('should handle single violation', () => {
+        transactionViolations = [{name: 'duplicatedTransaction', type: 'violation', userMessage: ''}];
+
+        const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
+        expect(result.value).toEqual(expect.arrayContaining(transactionViolations));
+    });
+
+    it('should handle multiple violations', () => {
+        transactionViolations = [
+            {name: 'duplicatedTransaction', type: 'violation', userMessage: ''},
+            {name: 'receiptRequired', type: 'violation', userMessage: ''},
+        ];
+        const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+        expect(result.value).toEqual(expect.arrayContaining(transactionViolations));
+    });
+
+    describe('policyRequiresCategories', () => {
+        beforeEach(() => {
+            policyRequiresCategories = true;
+            policyCategories = {Food: {enabled: true}};
+            transaction.category = 'Food';
+        });
+
+        it('should add missingCategory violation if no category is included', () => {
+            transaction.category = null;
+
+            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
+            expect(result.value).toEqual(expect.arrayContaining([missingCategoryViolation, ...transactionViolations]));
+        });
+
+        it('should add categoryOutOfPolicy violation when category is not in policy', () => {
+            transaction.category = 'Bananas';
+
+            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
+
```

Comment:
Ideally let's avoid adding blank lines that aren't preceded by comments
```suggestion
            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
            expect(result.value).toEqual(expect.arrayContaining([missingCategoryViolation, ...transactionViolations]));
        });

        it('should add categoryOutOfPolicy violation when category is not in policy', () => {
            transaction.category = 'Bananas';
            const result = ViolationsUtils.getViolationsOnyxData(transaction, transactionViolations, policyRequiresTags, policyTags, policyRequiresCategories, policyCategories);
```